src/analysis/ASTnodefunctions.ts

```diff
@@ -29,120 +27,218 @@ import { FieldWeight, TypeWeightObject, Variables } from '../@types/buildTypeWei
+            // check if the operation is in the type weights object.
+            if (node.operation.toLocaleLowerCase() in this.typeWeights) {
+                // if it is, it is an object type, add it's type weight to the total
+                complexity += this.typeWeights[node.operation].weight;
+                // console.log(`the weight of ${node.operation} is ${complexity}`);
+                // call the function to handle selection set node with selectionSet property if it is not undefined
+                if (node.selectionSet) {
+                    complexity += this.selectionSetNode(node.selectionSet, node.operation);
+                }
             }
+        } else if (node.kind === Kind.FRAGMENT_DEFINITION) {
+            // Fragments can only be defined on the root type.
+            // Parse the complexity of this fragment once and store it for use when analyzing other
+            // nodes. The complexity of a fragment can be added to the selection cost for the query.
+            const namedType = node.typeCondition.name.value;
+            // Duplicate fragment names are not allowed by the GraphQL spec and an error is thrown if used.
+            const fragmentName = node.name.value;
+
+            if (this.fragmentCache[fragmentName]) return this.fragmentCache[fragmentName];
+
+            const fragmentComplexity = this.selectionSetNode(
+                node.selectionSet,
+                namedType.toLowerCase()
+            );
+
+            this.fragmentCache[fragmentName] = fragmentComplexity;
+            return complexity; // Don't count complexity here. Only when fragment is used.
+        } else {
+            // TODO: Verify that are no other type definition nodes that need to be handled (see ast.d.ts in 'graphql')
+            // Other types include TypeSystemDefinitionNode (Schema, Type, Directvie) and
+            // TypeSystemExtensionNode(Schema, Type);
+            throw new Error(`ERROR: ASTParser.definitionNode: ${node.kind} type not supported`);
         }
+        return complexity;
     }
-    // TODO: add checks for Kind.FRAGMENT_DEFINITION here (there are other type definition nodes that i think we can ignore. see ast.d.ts in 'graphql')
-    return complexity;
-}
 
-export function documentNode(
-    node: DocumentNode,
-    typeWeights: TypeWeightObject,
-    variables: Variables
-): number {
-    let complexity = 0;
-    // iterate through 'definitions' array on the document node
-    for (let i = 0; i < node.definitions.length; i += 1) {
-        // call the function to handle the various types of definition nodes
-        complexity += definitionNode(node.definitions[i], typeWeights, variables);
+    documentNode(node: DocumentNode): number {
+        let complexity = 0;
+        // iterate through 'definitions' array on the document node
+        // FIXME: create a copy to preserve original AST order if needed elsewhere
+        const sortedDefinitions = [...node.definitions].sort((a, b) =>
```

Comment:
Any reason why we are sorting this here?

Reply:
I seen on line 237 you want the fragments parsed first, is the sorting accounting for that? It's still marked as a `fixme` on line 237 so it's not clear.